### PR TITLE
feat: XMSK-38 A simple IPC app

### PR DESCRIPTION
**Pull request by abdulla-ashurov**

### CODE REVIEW

File: interprocess-communication/parent_main.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include "file_mapping.hpp"
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:1]: Better divide libraries as a group: basic C++ libraries, Windows specific libraries, your libraries, e.g.:
```
#include <sstream>

#incude <Windows.h>

#include "file_mapping.hpp"
#include "process.hpp"
```

File: interprocess-communication/parent_main.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include "process.hpp"
+#include <sstream>
+#include <tchar.h>
+
+int main() {
+	try {
+		wchar_t msg[] = L"Hello World!";
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:8]: Using array of wchar_t type is low-level code or like C code. Instead of array of wchar_t type use std::string or std::wstring containers.

File: interprocess-communication/parent_main.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+		FileMapping f_map(255, FILE_MODES::WRITE, L"hFileMap", true);
+
+		CopyMemory(f_map.begin(), msg, (_tcslen(msg) * sizeof(TCHAR)));
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:11]: Incorrect calculate length of `wchar_t`. Understand the difference between `wchar _t` and `TCHAR`.

Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:11]: Do not use `CopyMemory` function. It is not safe.

File: interprocess-communication/parent_main.cpp [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#include "file_mapping.hpp"
+#include "process.hpp"
+#include <sstream>
+#include <tchar.h>
+
+int main() {
+	try {
+		wchar_t msg[] = L"Hello World!";
+		FileMapping f_map(255, FILE_MODES::WRITE, L"hFileMap", true);
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:9]: What's `255`, `L'hFileMap'`, `true` in `FileMapping` constructor. It seems like that hard to understand constructor of `FileMapping`.
1. Better to use 2 classes, those `FileMapping` and `InheritedFileMapping` for not using `true` argument in constructor.
2. Make `FileMapping` like class `std::array` that accepts size like non-template parameter for not using `255` in argument.
3. What's means FILE_MODES::WRITE? It means write/read or only write? It seems like for user it is not important what mode of `FileMapping` for this reason we can create `FileMapping` by default read/write for not using FILE_MODES in argument.
4. Do you create FileMapping's name in global or local scope? Also better to use this parameter in `InheritedFileMapping` constructor.

File: interprocess-communication/parent_main.cpp [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+		CopyMemory(f_map.begin(), msg, (_tcslen(msg) * sizeof(TCHAR)));
+
+		std::wcout << "Message: " << (const wchar_t*)(f_map.begin()) << std::endl;
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:13]: It better to add unit tests instead of prints debug messages.

File: interprocess-communication/parent_main.cpp [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+		const std::wstring app_name(L"C:\\Users\\ashur\\source\\repos\\interprocess-communication\\x64\\Release\\interprocess-communication.exe ");
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:15]: Child process name and child process arguments should be passed like argument of `parent_main` binary. Also, I can pass one/multiple handles like arguments.

File: interprocess-communication/parent_main.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+		std::wstringstream cmd;
+		cmd << app_name << (UINT_PTR)(f_map.handle());
+
+		Process process(cmd.str(), true);
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:19]: Child process name and child process arguments should be passed like argument of `parent_main` binary. Also, I can pass one/multiple handles like arguments.

Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:19]: After constructor of `Process` it is stopped or are we waiting to it? It will be good to add 2 functions like `std::thread`:
1. `process.detach()`
2. `process.join()`
Using these are functions, we can understand above processes for `Process` class.

File: interprocess-communication/parent_main.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+	}
+	catch (...) {
```
Comment (abdulla-ashurov) [inline interprocess-communication/parent_main.cpp:22]: Better to use specific error exceptions/codes instead of any in catch.

File: interprocess-communication/child_main.cpp [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	if (argc < 2) {
+		std::wcout << "Please, provide FileMapping Handle to child process using CLI\n";
+		return 1;
+	}
+
+	HANDLE hFileMap = (HANDLE)(std::stoi(argv[1]));
+	std::wcout << "hFileMap: " << hFileMap << std::endl;
+
+	FileMapping f_map(255, hFileMap, FILE_MODES::WRITE);
+	std::wcout << "Message from first process: " << (const wchar_t*)(f_map.begin());
```
Comment (abdulla-ashurov) [inline interprocess-communication/child_main.cpp:13]: It better to add unit tests instead of prints debug messages.

File: interprocess-communication/file_mapping.hpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
+#ifndef __FILE_MAPPING_HPP__
+#define __FILE_MAPPING_HPP__
+
+#include <iostream>
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:4]: Better divide libraries as a group: basic C++ libraries, Windows specific libraries, your libraries, e.g.:
```
#include <sstream>

#incude <Windows.h>

#include "file_mapping.hpp"
#include "process.hpp"
```

File: interprocess-communication/file_mapping.hpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+#include <string>
+#include "Windows.h"
+#include "common.hpp"
+
+enum FILE_MODES {
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:9]: Remove `FILE_MODES`. By default, create `FileMapping` in `read/write` mode.

File: interprocess-communication/file_mapping.hpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	HANDLE m_hFileMap;
+	void* m_buffer;
+
+	void free_members() {
+		if (m_buffer) {
+			UnmapViewOfFile(m_buffer);
+		}
+
+		if (m_hFileMap) {
+			CloseHandle(m_hFileMap);
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:27]: It better use `std::unique_ptr<HANDLE> m_hFileMap` instead of `CloseHandle()`.

File: interprocess-communication/file_mapping.hpp [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+class FileMapping {
+private:
+	size_t m_size;
+	std::wstring m_name;
+	HANDLE m_hFileMap;
+	void* m_buffer;
+
+	void free_members() {
+		if (m_buffer) {
+			UnmapViewOfFile(m_buffer);
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:23]: It better to use `std::unique_ptr<void*> m_buffer` instead of `UnmapViewOfFile(m_buffer)`.

File: interprocess-communication/file_mapping.hpp [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+enum FILE_MODES {
+	READ = PAGE_READONLY,
+	WRITE = PAGE_READWRITE
+};
+
+class FileMapping {
+private:
+	size_t m_size;
+	std::wstring m_name;
+	HANDLE m_hFileMap;
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:18]: It better to use `FileMappingHandle` class instead of `m_size`, `m_name`, `m_hFileMap`, `buffer`.

File: interprocess-communication/file_mapping.hpp [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+		if (m_buffer) {
+			UnmapViewOfFile(m_buffer);
+		}
+
+		if (m_hFileMap) {
+			CloseHandle(m_hFileMap);
+		}
+	}
+
+	void map_view_of_file(const size_t size, const FILE_MODES& mode) {
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:31]: Move `map_view_of_file` to `common.hpp`.

File: interprocess-communication/file_mapping.hpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+			free_members();
+			throw(std::runtime_error("Cannot map view of file!"));
+		}
+	}
+
+public:
+	FileMapping(const size_t size, const FILE_MODES& mode, const std::wstring& name = L"", const bool inheritHandle = false)
+		: m_size(size), m_name(name), m_hFileMap(NULL), m_buffer(NULL) {
+		SECURITY_ATTRIBUTES sa = create_security_attr(inheritHandle);
+		m_hFileMap = CreateFileMapping(INVALID_HANDLE_VALUE, &sa, mode, 0, m_size, m_name.length() ? m_name.c_str() : NULL);
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:51]: Wrap `CreateFileMapping` function also likes `MapViewOfFile`.

File: interprocess-communication/file_mapping.hpp [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+		if (m_hFileMap == NULL) {
+			throw(std::runtime_error("Cannot create file mapping object!"));
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:53]: It better to use specific error exceptions/codes instead of any in catch.

File: interprocess-communication/file_mapping.hpp [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+
+		map_view_of_file(m_size, mode);
+	}
+
+	FileMapping(const size_t size, HANDLE hFileMap, const FILE_MODES& mode)
+		: m_size(size), m_hFileMap(hFileMap), m_buffer(NULL) {
+		map_view_of_file(size, mode);
+	}
+
+	FileMapping(const FileMapping& other) = delete;
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:64]: Implement `FileMapping` copy constructor.

File: interprocess-communication/file_mapping.hpp [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+	FileMapping& operator=(const FileMapping& other) = delete;
+
+	size_t size() const {
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:68]: Move all these function to `FileMappingHandle` class.

File: interprocess-communication/file_mapping.hpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	void* end() {
+		return static_cast<uint8_t*>(m_buffer) + m_size;
+	}
+
+	HANDLE handle() {
+		return m_hFileMap;
+	}
+
+	~FileMapping() {
```
Comment (abdulla-ashurov) [inline interprocess-communication/file_mapping.hpp:86]: It better to use style where you sort functions by importance, e.g.: constructor, copy constructor, operator=, destructor. Also, you can use member protectors to divide different sections.

File: interprocess-communication/process.hpp [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+	Process(const std::wstring& cmd, const bool inherit_handles = false) : m_pi{ 0 } {
+		SECURITY_ATTRIBUTES sa = create_security_attr(inherit_handles);
+		STARTUPINFO si = { sizeof(si) };
+
+		if (CreateProcess(NULL, const_cast<LPWSTR>(cmd.c_str()), &sa, &sa, inherit_handles, NULL, NULL, NULL, &si, &m_pi) == NULL) {
+			std::wcout << "It cannot create a new process" << std::endl;
+			return;
+		}
+
+		WaitForSingleObject(m_pi.hProcess, INFINITE);
```
Comment (abdulla-ashurov) [inline interprocess-communication/process.hpp:23]: It is a bad idea to putting WaitForSingleObject in constructor because let's say if I use it std::thread class then thread will not start until constructor of Process is not created fully.
Move it to destructor of `Process` class.

File: interprocess-communication/parent_main.cpp (hunk 1 above)
Comment (abdulla-ashurov) [reply interprocess-communication/parent_main.cpp]: Done!
